src/overlays/actors/ovl_En_Racedog/z_en_racedog.c

```diff
@@ -628,20 +628,41 @@ s32 EnRacedog_IsOverFinishLine(EnRacedog* this, Vec2f* finishLineCoordinates) {
 
     // frontPointsCrossProduct is positive if the dog is to the left of the line formed by the front points
     // crossProductTemp is positive if the dog is above the line formed by the bottom points
+    // This is checking that the dog within the region defined by front and bottom lines like so:
+    //         |
+    //    X    | Front
+    //         |
+    // --------
+    //  Bottom
     frontPointsCrossProduct = ((xDistToTopFront * zDistToBottomFront) - (xDistToBottomFront * zDistToTopFront));
     crossProductTemp = (((xDistToBottomFront * zDistToBottomBack) - (xDistToBottomBack * zDistToBottomFront)));
+    //! @bug If the dog is *precisely* on top of the line formed by the front points, then frontPointsCrossProduct
+    //! will be zero. This will cause this multiplication (and all future multiplications) to be zero, which will
+    //! make this function think the dog has crossed the finish line. The line formed by the front points extends
+    //! throughout the entire racetrack, so a dog can trigger this when they're not even close to the actual finish
+    //! line, causing them to finish the race incredibly early.
     if (frontPointsCrossProduct * crossProductTemp < 0.0f) {
         return false;
     }
```

Comment:
Is this 0.0f to floating point precision? What if `frontPointsCrossProduct` is 1e-6 for example?

Reply:
Well, I set `frontPointsCrossProduct` to `0.00000000000000000000000000001f` if the dog is the selected dog, and the bug I highlighted in the description doesn't occur. Sometimes, a different bug manifests, where a dog can win a race in the final run up to the finish line, but a decent bit before they truly cross it:

https://github.com/zeldaret/mm/assets/12245827/a3cc5912-6cd4-4de7-92ee-bab491007d9b

But in this case, this is *not* on the line segment created by the front two points, so something else is going on. It might be that the other parts of the multiplication become small enough that floating point imprecision rounds them down to 0?

If I instead set the `frontPointsCrossProduct` to `0.0f`, my dog instantly wins as soon as it is able to do so (remember that EnRacedog has a bit of anticheat in it, so I can't win immediately). Which is proof that my code changes are working, that my understanding of this is correct, and that somehow `0.00000000000000000000000000001f` is too big of a value.

I don't exactly know how to test if it's floating-point precise, but if it's not, it's incredibly close.

Reply:
Actually, the first video I showed in the prior post isn't a realistic bug. In normal gameplay, the `frontPointsCrossProduct` is *not* going to be a very tiny positive number there. In fact, it will be a decent-sized negative number, meaning that the result you see there is just totally faked and could never possibly happen without modifying the game. What we really need to test is "when the dog is close to the line formed by the front points, *how* close do they have to be to trigger it?" So I wrote some code like this:
```c
    frontPointsCrossProduct = ((xDistToTopFront * zDistToBottomFront) - (xDistToBottomFront * zDistToTopFront));
    if (this->index == this->selectedDogIndex && ABS(frontPointsCrossProduct) < 10000.0f) {
        frontPointsCrossProduct = 0.0000000000000000000000000000000001f;
    }
```

With this code, the dog never finished early. When I changed `0.0000000000000000000000000000000001f` to `0.0f`, I got the *exact* same behavior as in the twitch clip; I won the race when I hit the line formed by the front points on the other side of the track.

So I think it's safe to conclude that, if this is *not* floating-point precise, it is incredibly close to being so, to the point where it may be a distinction without a difference.

Reply:
You can see here in discord that I managed to force the minimum possible normalized positive floating point value: https://discord.com/channels/688807550715560050/688851976746041391/1118293183865040937

And what happens? The bug doesn't occur:
https://github.com/zeldaret/mm/assets/12245827/83830af8-8542-488d-9830-e7deb248b426

In other words, this *is* floating-point precise. Here is my modified `EnRacedog_IsOverFinishLine` if you want to test for yourself. I'm saving the result to `sFrontPointsCrossProduct ` so I can print it later; it's not necessary as part of the actual computation:
```c
s32 EnRacedog_IsOverFinishLine(EnRacedog* this, Vec2f* finishLineCoordinates) {
    f32 xDistToTopFront;
    f32 zDistToTopFront;
    f32 xDistToBottomFront;
    f32 xDistToBottomBack;
    f32 zDistToBottomBack;
    f32 zDistToBottomFront;
    f32 zDistToTopBack;
    f32 xDistToTopBack;
    f32 frontPointsCrossProduct;
    f32 crossProductTemp;
    s32 smallestFloatInHex = 0x00800000; // smallest normalized positive float in hex
    s32* intPointer = &smallestFloatInHex;
    f32* floatPointer = intPointer;
    f32 smallestFloat = *floatPointer;

    xDistToTopFront = this->actor.world.pos.x - finishLineCoordinates[0].x;
    zDistToTopFront = this->actor.world.pos.z - finishLineCoordinates[0].z;
    xDistToBottomFront = this->actor.world.pos.x - finishLineCoordinates[1].x;
    zDistToBottomFront = this->actor.world.pos.z - finishLineCoordinates[1].z;
    xDistToBottomBack = this->actor.world.pos.x - finishLineCoordinates[2].x;
    zDistToBottomBack = this->actor.world.pos.z - finishLineCoordinates[2].z;
    xDistToTopBack = this->actor.world.pos.x - finishLineCoordinates[3].x;
    zDistToTopBack = this->actor.world.pos.z - finishLineCoordinates[3].z;

    frontPointsCrossProduct = ((xDistToTopFront * zDistToBottomFront) - (xDistToBottomFront * zDistToTopFront));
    if (this->index == this->selectedDogIndex) {
        sFrontPointsCrossProduct = frontPointsCrossProduct;
        if (ABS(frontPointsCrossProduct) < 10000.0f) {
            Actor_PlaySfx(&this->actor, NA_SE_EN_LAST3_VOICE_DAMAGE2_OLD);
            frontPointsCrossProduct = smallestFloat;
        }
    }
    crossProductTemp = (((xDistToBottomFront * zDistToBottomBack) - (xDistToBottomBack * zDistToBottomFront)));
    if (frontPointsCrossProduct * crossProductTemp < 0.0f) {
        return false;
    }

    frontPointsCrossProduct = ((xDistToTopFront * zDistToBottomFront) - (xDistToBottomFront * zDistToTopFront));
    if (this->index == this->selectedDogIndex && ABS(frontPointsCrossProduct) < 10000.0f) {
        frontPointsCrossProduct = smallestFloat;
    }
    crossProductTemp = ((xDistToBottomBack * zDistToTopBack) - (xDistToTopBack * zDistToBottomBack));
    if (frontPointsCrossProduct * crossProductTemp < 0.0f) {
        return false;
    }

    frontPointsCrossProduct = ((xDistToTopFront * zDistToBottomFront) - (xDistToBottomFront * zDistToTopFront));
    if (this->index == this->selectedDogIndex && ABS(frontPointsCrossProduct) < 10000.0f) {
        frontPointsCrossProduct = smallestFloat;
    }
    crossProductTemp = ((xDistToTopBack * zDistToTopFront) - (xDistToTopFront * zDistToTopBack));
    if (frontPointsCrossProduct * crossProductTemp < 0.0f) {
        return false;
    }

    return true;
}
```

Reply:
Great, thanks for looking into it in so much detail. The two things I'd like to see in the comment is that 1) this is floating-point precise and that this can happen to any dog. Maybe:
`@bug If any dog is  on top of the line to floating-point precision, `
Perhaps? Or maybe there's a better way to phrase it

Reply:
I did a parenthetical statement like so:
```c
//! @bug If any dog is precisely (with floating-point precision) on top of the line formed by the front points,
```
What do you think?